### DISCUSSION TITLE
rebuildAndRecalculate hangs

### DISCUSSION THREAD

**Discussion by doczoidberg** [handsontable/hyperformula#544] [General]
What can I do when `rebuildAndRecalculate` freezes my browser?

On a specific sheetset I freezes forever.

**Comment by doczoidberg** [comment]
Are there also ways to speed the rebuild up or a way to avoid it?

It takes up several seconds for each sheetset for us and we have multiple sheetsets to calculate.

**Comment by bardek8** [comment]
Indeed, this function will slow down the engine when called too often.

In #493 I listed different solutions for the problem with REFs due to cross-sheet references. Maybe you can avoid calling this function at all? 
a) _First add/rename all sheets and then fill them_ - this should be the most efficient.
b) If you really need to call `rebuildAndRecalculate`, this should be enough to call it once for all sheets, you don't have to call it for each of them separately.

In principle, HyperFormula is designed in such a way that one doesn't have to call `rebuildAndRecalculate` at all. If the above-mentioned solutions do not work, maybe you have found a bug and we'll be happy to investigate it if you provide us with a small example.

**Comment by doczoidberg** [comment]
I do:
```
   hf.addSheet(name);
   hf.setSheetContent(name, cells);
```

after that I have `#ref` errors which I don't have after rebuilding all sheets. Do I something wrong?

**Comment by bardek8** [comment]
Yes, you should first add all sheets and only then set cell contents:
```
hf.addSheet(name1);
hf.addSheet(name2);
....

hf.setSheetContent(name1, cells);
hf.setSheetContent(name2, cells);
...
```
This will avoid the problem with references to non-existing sheets.

**Comment by doczoidberg** [comment]
ah ok, I'll try that

**Comment by doczoidberg** [comment]
it works without `rebuildandcalucate()` but also setting the sheet content takes a long time. Just loading a state would be nice https://github.com/handsontable/hyperformula/issues/545
